### PR TITLE
Refactor videotoclass

### PR DESCRIPTION
**Pull request by sesevgen**
Addresses #58 

Unit tests for both problems work, but I haven't tested a model.

### CODE REVIEW

**Comment by tkornut** [comment]
This pull request **introduces 1 alert** when merging d4e35b3977c4e1d7ee93eef4e4264351e9b21c53 into 194cedec3431c7b56d1c1b2fed1a6e75e9fec265 - [view on LGTM.com](https://lgtm.com/projects/g/IBM/mi-prometheus/rev/pr-54270c198f10de2bd8ec0ed45d5e69c2b8d89104)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py [hunk 1]
```diff
@@ -60,9 +60,9 @@ def __init__(self, params):
 
                 - ``self.data_definitions`` :
 
-                    >>> self.data_definitions = {'images': {'size': [-1, 3, -1, -1], 'type': [torch.Tensor]},
-                    >>>                          'mask': {'size': [-1, -1, -1, -1], 'type': [torch.Tensor]},
-                    >>>                          'targets': {'size': [-1, 1], 'type': [torch.Tensor]},
+                    >>> self.data_definitions = {'images': {'size': [-1, 28, 1, 1, 28], 'type': [torch.Tensor]},
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: Good catch! 👍

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py [hunk 2]
```diff
@@ -60,9 +60,9 @@ def __init__(self, params):
...
+                    >>>                          'mask': {'size': [-1, 28, 1], 'type': [torch.Tensor]},
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: So in here the dimensions are BATCH (-1) by SEQ_LEN (28) by OUTPUT DIM (1)

Good! :)

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py [hunk 3]
```diff
@@ -132,13 +132,13 @@ def __getitem__(self, index):
...
 
         # create mask
-        mask = torch.zeros(self.num_rows).type(self.app_state.IntTensor)
-        mask[-1] = 1
+        mask = torch.zeros((self.num_rows,1)).type(self.app_state.IntTensor)
+        mask[-1,0] = 1
 
         data_dict = DataDict({key: None for key in self.data_definitions.keys()})
-        data_dict['images'] = img
+        data_dict['images'] = img.view(28,1,1,28)
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: 👍

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py [hunk 4]
```diff
@@ -132,13 +132,13 @@ def __getitem__(self, index):
...
         data_dict['mask'] = mask
-        data_dict['targets'] = target
+        data_dict['targets'] = target*torch.ones((28,1),dtype=torch.long)
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: Ok, so the trick is that you are using here torch.long, whereas mask is of type: self.app_state.IntTensor

Generally self.app_state.IntTensor is a remainer of our previous solution, where app_state stored cpu/gpu types, depending on the setting. This appeared to be problematic when we started to use DataLoaders with multiple workers - we had issues with different loader workers tried to load different images to GPU directly.

So instead we decided that: getitem/collate return items/batches that remain in CPU memory, then the mip-worker transports them to GPU (DataDict has a function that does that automatically).

Solution: please replace all references to self.app_state.TYPE with torch.TYPE.

@vmarois Can you please confirm that?

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py [hunk 5]
```diff
@@ -180,6 +180,20 @@ def collate_fn(self, batch):
     # get a sample
     sample = problem[0]
     print(repr(sample))
+
+    # test whether data structures match expected definitions
+    # images should be (batch size x sequence x channel x height x width)
+    # as this is a sample, we should have (sequence x channel x height x width) == (28, 1, 1, 28)
+    assert sample['images'].shape == torch.ones((28, 1, 1, 28)).shape, "Unit test failed! Expected images shape {} but got {}".format(torch.ones((28*28, 1, 1, 1)).shape, sample['images'].shape)
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: Uuuu, nice! So we are really getting closer to something that is commonly understood as unit test ;)

File: miprometheus/problems/video_to_class/seq_mnist_to_class/sequential_pixel_mnist.py [hunk 6]
```diff
@@ -58,11 +58,11 @@ def __init__(self, params):
 
                 - ``self.data_definitions`` :
 
-                    >>> self.data_definitions = {'sequences': {'size': [-1, 28*28], 'type': [torch.Tensor]},
-                    >>>                          'mask': {'size': [-1, 28*28], 'type': [torch.Tensor]},
-                    >>>                          'targets': {'size': [-1], 'type': [torch.Tensor]},
-                    >>>                          'targets_label': {'size': [-1, 1], 'type': [list, str]}
-                    >>>                         }
+                    >>> self.data_definitions = {'images': {'size': [-1, 28*28, 1, 1, 1], 'type': [torch.Tensor]},
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/sequential_pixel_mnist.py]: 👍

File: miprometheus/problems/video_to_class/seq_mnist_to_class/sequential_pixel_mnist.py [hunk 7]
```diff
@@ -124,13 +124,13 @@ def __getitem__(self, index):
         label = self.labels[target.data]
 
         # create mask
-        mask = torch.zeros(self.num_rows * self.num_columns).type(self.app_state.IntTensor)
-        mask[-1] = 1
+        mask = torch.zeros((self.num_rows * self.num_columns,1)).type(self.app_state.IntTensor)
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/sequential_pixel_mnist.py]: Similarly: use torch.IntTensor

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py [hunk 8]
```diff
@@ -132,13 +132,13 @@ def __getitem__(self, index):
         label = self.labels[target.data]
 
         # create mask
-        mask = torch.zeros(self.num_rows).type(self.app_state.IntTensor)
-        mask[-1] = 1
+        mask = torch.zeros((self.num_rows,1)).type(self.app_state.IntTensor)
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: Please use torch.IntTensor instead

File: miprometheus/problems/video_to_class/video_to_class_problem.py [hunk 9]
```diff
@@ -41,9 +41,9 @@ def __init__(self, params):
             - Sets loss function to ``CrossEntropy``,
             - sets ``self.data_definitions`` to:
 
-                >>> self.data_definitions = {'images': {'size': [-1, 3, -1, -1], 'type': [torch.Tensor]},
-                >>>                          'mask': {'size': [-1, -1, -1, -1], 'type': [torch.Tensor]},
-                >>>                          'targets': {'size': [-1, 1], 'type': [torch.Tensor]},
+                >>> self.data_definitions = {'images': {'size': [-1, -1, 3, -1, -1], 'type': [torch.Tensor]},
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/video_to_class_problem.py]: 👍

File: miprometheus/problems/video_to_class/video_to_class_problem.py [hunk 10]
```diff
@@ -140,7 +144,7 @@ def collect_statistics(self, stat_col, data_dict, logits):
         """
         stat_col['acc'] = self.calculate_accuracy(data_dict, logits)
 
-    def show_sample(self, data_dict, sample_number=0):
+    def show_sample(self, data_dict, sample_number=0, sequence_number=0):
```
Comment (tkornut) [inline miprometheus/problems/video_to_class/video_to_class_problem.py]: Makes sense!

Comment (tkornut) [reply miprometheus/problems/video_to_class/video_to_class_problem.py]: Can you please update the function description?

File: miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py (hunk 4 above)
Comment (vmarois) [reply miprometheus/problems/video_to_class/seq_mnist_to_class/permuted_sequential_row_mnist.py]: Yes, I confirm this indeed. This is the line which converts the cpu tensor types to CUDA ones:

https://github.com/IBM/mi-prometheus/blob/master/miprometheus/workers/worker.py#L474

So batch generation stays on cpu for now, but we might want to change that in the future if we use [PyTorch's DataParallel](https://pytorch.org/docs/stable/nn.html?highlight=dataparallel#torch.nn.DataParallel) object

File: miprometheus/problems/video_to_class/video_to_class_problem.py [hunk 11]
```diff
@@ -154,13 +158,21 @@ def show_sample(self, data_dict, sample_number=0):
         import matplotlib.pyplot as plt
 
         # Unpack dict.
-        images, masks, targets, labels = data_dict.values()
+        #images, masks, targets, labels = data_dict.values()
+        images = data_dict['images']
```
Comment (vmarois) [inline miprometheus/problems/video_to_class/video_to_class_problem.py]: Nice 👍

**Comment by sesevgen** [comment]
Addressed changes requested above.

**Comment by tkornut** [comment]
Fixed #58
